libnmstate/metadata.py

```diff
@@ -181,3 +202,127 @@ def _generate_route_metadata(desired_state):
+    """
+    Save DNS configuration to choose interface as metadata.
+    Raise NmstateValueError if failed to find interface.
+    To handle 3 DNS name server, we need 2+ interfaces when IPv6 server been
+    placed between two IPv4 servers and it also require assigning DNS priority.
+    To simplify workflow, currently we only support at most 2 name servers.
+    """
+    servers = desired_state.config_dns.get(DNS.SERVER, [])
+    searches = desired_state.config_dns.get(DNS.SEARCH, [])
+    if len(servers) > 2:
+        raise NmstateNotImplementedError(
+            'Nmstate only support at most 2 DNS name servers')
+    if not servers and not searches:
+        return
+    ipv4_server = []
+    ipv6_server = []
+    for server in servers:
+        if iplib.is_ipv6_address(server):
+            ipv6_server.append(server)
+        else:
+            ipv4_server.append(server)
+
+    search_saved = False
+    ipv4_iface, ipv6_iface = _choose_iface_for_dns(desired_state,
+                                                   current_state)
+
+    # Set DNS priority in case IPv4 server is listed before IPv6 server
+    need_dns_priority = False
+    if ipv6_server and ipv4_server:
+        if servers == [ipv4_server[0], ipv6_server[0]]:
+            need_dns_priority = True
+
+    for iface, family, servers in ((ipv6_iface, Interface.IPV6, ipv6_server),
+                                   (ipv4_iface, Interface.IPV4, ipv4_server)):
+        if servers:
+            if not iface:
+                raise NmstateValueError(
+                    'Failed to find suitable interface for saving DNS '
+                    'name servers: %s' % servers)
+            if iface not in desired_state.interfaces:
+                desired_state.include_changed_interfaces(current_state, [iface])
+            dns_meta = {
+                DNS_PRIORITY: DNS_DEFAULT_PRIORITY_FOR_STATIC,
+                DNS.SERVER: servers,
+                DNS.SEARCH: []
+            }
+            if need_dns_priority:
+                if family == Interface.IPV6:
+                    dns_meta[DNS_PRIORITY] = DNS_ORDERING_IPV6_PRIORITY
+                else:
+                    dns_meta[DNS_PRIORITY] = DNS_ORDERING_IPV4_PRIORITY
+            if not search_saved:
+                dns_meta[DNS.SEARCH] = searches
+                search_saved = True
+            iface_state = desired_state.interfaces[iface]
+            if family not in iface_state:
+                iface_state[family] = {DNS_METADATA: dns_meta}
+            else:
+                iface_state[family][DNS_METADATA] = dns_meta
+    # Some interfaces might old DNS configuration which is not included
+    # in desired_state, we need to add them into desired_state in order to
+    # clear its DNS configurations
+    old_dns_ifaces = nm_dns.get_dns_config_iface_names()
+    desired_state.include_changed_interfaces(current_state, old_dns_ifaces)
+
+
+def _choose_iface_for_dns(desired_state, current_state):
```

Comment:
I suggest you move the lookup logic outside the metadata, probably fits better under the nm.dns, being a NM specific limitation.